### ISSUE TITLE
Research: Performance Enhancements for Time out Issues

### ISSUE THREAD

**Issue by roamye** [project-lux/lux-marklogic#136]
**Problem Description:**
There have been several time out errors in Advanced Search queries which only have one multiple field. (see example below) 
This issue serves as a research ticket for all performance enhancements for time out issues not related to #113 to figure out how to solve this issue. 

Another time out issue is when users select the last page of a search result. (example below) 

**Expected Behavior/Solution:**
Research on how to fix this time out issue. Solution is TBD
Possible solution: increase our time out

**Requirements:**
TBD

**Needed for promotion:**
If an item on the list is not needed, it should be crossed off but not removed.

- [ ] Wireframe/Mockup - Mike
- [ ] Committee discussions - Sarah
- [ ] Feasibility/Team discussion - Sarah
- [ ] Backend requirements - TBD
- [ ] Frontend requirements- TBD
- [ ] Questions
- List of questions for discussions. Answers should be documented within the issue.

**UAT/LUX Examples:**

- AS query [timeout](https://lux-front-tst.collections.yale.edu/view/results/works?facetRequest=true&q=%7B%22AND%22%3A%5B%7B%22classification%22%3A%7B%22name%22%3A%22archive%22%7D%7D%5D%7D&view=grid&wf=%7B%22AND%22%3A%5B%7B%22classification%22%3A%7B%22id%22%3A%22https%3A%2F%2Flux.collections.yale.edu%2Fdata%2Fconcept%2Fdd625870-8775-44b2-9c19-b6281462c684%22%7D%7D%5D%7D&wp=14929)
- selecting last page on SS for colors: https://lux-front-tst.collections.yale.edu/view/results/objects?q=%7B%22text%22%3A%22colors%22%2C%22_lang%22%3A%22en%22%7D&sq=colors&view=list&is=anySortName%3Aasc&ip=5457

**Dependencies/Blocks:**

- Blocked By: Issues that are blocking the completion of the current issue.
- Blocking: Issues being blocked by the completion of the current issue.

**Related Github Issues:**

- #223 is a way to quickly return any page of from a large search results page.

**Related links:**
- Bugherd: https://www.bugherd.com/projects/284041/tasks/2180

**Wireframe/Mockup:**
Place wireframe/mockup for the proposed solution at end of ticket.

**Comment by brent-hartwig** [comment]
This ticket's scope may be too broad.  It's typically better to focus on one timeout / perf issue at a time.

In the case of requesting the last page of a search's results, the frontend is specifying pagination parameters that has the current implementation of the backend perform a search that has to _filter_ results through page * page length.  Filtering is the process of pulling the document from disk and through d-node then e-node caches to validate it meets all search criteria.  The process _may_ drop that were determined exclusively via indexes (unfiltered).

When searching for objects containing "colors", there are 109,127.  With cold caches, it 22.5 seconds to get to the last seven.  With warm caches, it still takes 16.1 seconds.  An unfiltered search consistently returns the same last seven results in 117 milliseconds, illustrating this search can be accurately resolved by indexes alone and that the filtering process is adding 16 or seconds to sift through 110K results.

Things we may want to discuss:

1. Switching to unfiltered searches and seeking out instances where false positives show up.
2. When the requested page is more than 50% into cts.estimate, reverse the sort and paginate/filter the other way.
3. Do not offer a Last page link in the frontend.  What value does it offer?  Reverse the sort when interested in those results.

**Comment by brent-hartwig** [comment]
I received some internal feedback:

> 16 sec for 100k is 0.16 msec each doc which is pretty quick. Maybe it isn't expected that it is single threaded on the enode but, I believe that is the case for that stage of the query processing.

I confirmed the single-thread belief and submitted https://progressdataplatform.ideas.aha.io/ideas/ML-I-42.

Having said that, I am still much more in favor option no. 1.

**Comment by clarkepeterf** [comment]
@brent-hartwig I agree it would be great to look at option no. 1 - Many of our searches are resolvable only using indexes so should be able to go unfiltered. Would be interesting to see which subset of searches requires filtering and if there is a way to enable it to work unfiltered or just accept the false positives

**Comment by brent-hartwig** [comment]
> Would be interesting to see which subset of searches requires filtering and if there is a way to enable it to work unfiltered or just accept the false positives

Part of CTS/Optic batch no. 3 👍 .  We are becoming familiar with what cannot be resolved by indexes, including punctuation and whitespace.  We need a complete list and decide whether those are important enough to LUX to accept the performance penalty imposed by filtered and prevent us from adopting Optic.  Whenever we encounter a false positive that isn't explained by the list, we need to question our index configuration and code.

**Comment by brent-hartwig** [comment]
@jffcamp, this ticket's example of timing out when requesting the last page of a large search results set could be addressed by unfiltered searches (#223).  Here's the 3rd para from [my 9 May comment](https://github.com/project-lux/lux-marklogic/issues/136#issuecomment-2103398285):

> When searching for objects containing "colors", the estimate (unfiltered) comes back 109,127. With cold caches, it took 22.5 seconds to get to the last seven. With warm caches, it still took 16.1 seconds. An unfiltered search consistently returns the same last seven results in 117 milliseconds, illustrating this search can be accurately resolved by indexes alone and that the filtering process is adding 16 or seconds to double check the 109K results.

**Comment by roamye** [comment]
Left a note on #223, leaving it below this issue for reference. 
https://github.com/project-lux/lux-marklogic/issues/223#issuecomment-2293663758

**Comment by prowns** [comment]
@roamye and @clarkepeterf - I cannot replicate this.  propose close?

**Comment by clarkepeterf** [comment]
Same here. Fine with me to close

**Comment by roamye** [comment]
Closing per ticket forming
